### DISCUSSION TITLE
Get a non-const Function

### DISCUSSION THREAD

**Discussion by xuxiaobei1995** [idaholab/moose#17796] [Q&A General] [answered]
Hi all,

Recently I updated my MOOSE and found that the returned type of getFunction() has been changed from "Function &" to "const Function &". However in my code I want to get the function and modify its data, which means a non-const Function is needed. Does anyone know the way to realize this? Any suggestion is appreciated.

Xiaobei

**Comment by WilkAndy** [comment]
See the accepted answer instead of this one.

Have a look online, or inside MOOSE, for `const_cast`

**Reply by xuxiaobei1995** [reply]
Thank you Andy! This is just what I need.

**Comment by aeslaughter** [comment, marked as answer]
I wouldn't recommend this, it is `const` for a reason. Any object can get a function, thus if it is non-const then any object could change it. You will find yourself in a situation that by adding a Material that uses a function changes the behavior of the function that is also used by some other object.

**Reply by hugary1995** [reply]
I second @aeslaughter 's point. @xuxiaobei1995 if you can let us know what you want to do by "modifying its data", we can offer more help.

**Reply by xuxiaobei1995** [reply]
Thanks for your suggestion! What I want is to give an array to my function. The function is computed based on the array. I added a vector as a data member in the function for storing the array. My previous way is to get the function (using getFunction()) and access the vector directly. Do you know the safer way to implement this? @aeslaughter @hugary1995

**Reply by aeslaughter** [reply]
What type of object is giving the array to the Function? And when do you need this to happen?

**Reply by loganharbour** [reply]
You should probably add a method that takes a const view of your array. 

Functions really should only be used for simple calls when you always pass a position and time to get a value. I would advise against adding methods to them, because in my opinion you should only be using a Function when it follows through the normal value interfaces that take a point and time.

I would suggest creating a UserObject that has such a custom method on it (where you pass in your data). You can then couple to this UserObject and call its methods from within your other MooseObject.

**Reply by xuxiaobei1995** [reply]
> What type of object is giving the array to the Function? And when do you need this to happen?

It is a transfer which receives the array from the masterapp. So it happens at TIMESTEP_BEGIN or TIMESTEP_END

**Reply by xuxiaobei1995** [reply]
> You should probably add a method that takes a const view of your array.
> 
> Functions really should only be used for simple calls when you always pass a position and time to get a value. I would advise against adding methods to them, because in my opinion you should only be using a Function when it follows through the normal value interfaces that take a point and time.
> 
> I would suggest creating a UserObject that has such a custom method on it (where you pass in your data). You can then couple to this UserObject and call its methods from within your other MooseObject.

Thanks for your nice advice! It is indeed more reasonable to store the array in a Userobject and let the function couples the userobject to get the data.

**Reply by loganharbour** [reply]
> It is indeed more reasonable to store the array in a Userobject and let the function couples the userobject to get the data.

I don't think you want to _store_ the array in the UserObject. Let's say your data structure that you're passing is a `std::vector<Real>`, my suggestion is to create a member function on your UO that has a signature something like:

```
Real getValue(const std::vector<Real> & data) const;
```

In this case, you're passing a const reference to the UO so that it can look at the data, and then create a value based on it. I'm not suggesting that you actually _store_ the data (your array) in the UO. By doing it this way, you only need const access to the UO and can call const methods on it.

**Reply by xuxiaobei1995** [reply]
@loganharbour Thank you Logan! I think I know how to do it now.
